src/measurement/position_measurement.js

```diff
@@ -446,66 +452,106 @@ export function wrappedLineExtentChar(cm, lineObj, preparedMeasure, target) {
+}
+
 function coordsCharInner(cm, lineObj, lineNo, x, y) {
+  // Move y into line-local coordinate space
   y -= heightAtLine(lineObj)
-  let begin = 0, end = lineObj.text.length
   let preparedMeasure = prepareMeasureForLine(cm, lineObj)
-  let pos
+  // When directly calling `measureCharPrepared`, we have to adjust
+  // for the widgets at this line.
+  let widgetHeight = widgetTopHeight(lineObj)
+  let begin = 0, end = lineObj.text.length, ltr = true
+
   let order = getOrder(lineObj, cm.doc.direction)
+  // If the line isn't plain left-to-right text, first figure out
+  // which bidi section the coordinates fall into.
   if (order) {
-    if (cm.options.lineWrapping) {
-      ;({begin, end} = wrappedLineExtent(cm, lineObj, preparedMeasure, y))
-    }
-    pos = new Pos(lineNo, Math.floor(begin + (end - begin) / 2))
-    let beginLeft = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left
-    let dir = beginLeft < x ? 1 : -1
-    let prevDiff, diff = beginLeft - x, prevPos
-    let steps = Math.ceil((end - begin) / 4)
-    outer: do {
-      prevDiff = diff
-      prevPos = pos
-      // Make these steps don't take us outside of the current bidi part
-      let bidiPart = order[getBidiPartAt(order, pos.ch, dir < 0 ? "before" : "after")]
-      let maxSteps = Math.max(1, Math.min(steps, (dir < 0) != (bidiPart.level % 2 > 0) ? pos.ch - bidiPart.from : bidiPart.to - pos.ch))
-      for (let i = 0; i < maxSteps; ++i) {
-        let prevPos = pos
-        pos = moveVisually(cm, lineObj, pos, dir)
-        if (pos == null || pos.ch < begin || end <= (pos.sticky == "before" ? pos.ch - 1 : pos.ch)) {
-          pos = prevPos
-          break outer
-        }
-      }
-      diff = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left - x
-      if (steps > 1) {
-        let diffChangePerStep = Math.abs(diff - prevDiff) / steps
-        steps = Math.min(steps, Math.ceil(Math.abs(diff) / diffChangePerStep))
-        dir = diff < 0 ? 1 : -1
+    // Bidi parts are sorted left-to-right. Find the first part whose
+    // end is after the given coordinates.
+    let index = findFirst(i => {
+      let part = order[i], ltr = part.level % 2 == 0
+      return boxIsAfter(cursorCoords(cm, Pos(lineNo, ltr ? part.to : part.from, ltr ? "before" : "after"),
+                                     "line", lineObj, preparedMeasure), x, y, true)
+    }, 0, order.length - 1)
+    let part = order[index]
+    ltr = part.level % 2 == 0
+    // If this isn't the first part, the part's start is also after
+    // the coordinates, and the coordinates aren't on the same line as
+    // that start, move one part back.
+    if (index > 0) {
+      let start = cursorCoords(cm, Pos(lineNo, ltr ? part.from : part.to, ltr ? "after" : "before"),
+                               "line", lineObj, preparedMeasure)
+      if (boxIsAfter(start, x, y, true) && start.top > y) {
+        part = order[--index]
+        ltr = part.level % 2 == 0
```

Comment:
This is wrong in cases with lvl 2, I think. Example:

http://unicode.org/cldr/utility/bidi.jsp?a=abcd+%D9%85+1+%D8%B1&p=LTR&hack=on

If your line fits 6 characters, according to [UBA](http://unicode.org/reports/tr9/#Reordering_Resolved_Levels), it will be  
```
012345
------
abcd E 
N F
```

But the bidi ordering is:

```
"{"level":0,"from":0,"to":5}, {"level":1,"from":8,"to":10}, {"level":2,"from":7,"to":8}, {"level":1,"from":5,"to":7}"
```

The first bidi part in bidi ordering order that's after, say, the right end of `E` will be `order[1]`. I suppose you are looking for `order[3]` though?

Reply:
The unicode.org link isn't loading right now, so I'm not sure what you mean, but I guess you're saying that, in the presence of wrapped lines with level 2 spans in them, the visual ordering may not correspond to the ordering in `order`? Experiments seem to conform that. That's bad!

I'll see if I can work around this by doing the clipping to a single wrapped line _before_ finding the appropriate bidi part (discarding those outside of the line's range). I guess that makes applying a binary search a lot harder, but I suppose falling back to a full search in the wrapped case (and discarding all spans that aren't in the line) should be okay.

Reply:
Hm, it's loading for me. Anyway, what I am saying is: There are cases in which a span `A` that comes before a span `B` _in the bidi order_ comes after the span `B` _in top-down order_.